File: core/lua/jitaccess/core/crypto.lua
-- jitaccess.core.crypto — proof & nonce cryptography (docs/PROTOCOL.md §3, §5)
--
-- Depends on lua-resty-openssl (bundled with BunkerWeb/OpenResty) for HMAC-SHA256
-- and a CSPRNG, and on ngx.{encode,decode}_base64 for base64url.
--
-- NOTE: not executed on the dev host (no Lua runtime). The proof and nonce
-- constructions were cross-checked against core/testdata/vectors.json via an
-- equivalent port before commit; the harness re-validates under real Lua.

local hmac    = require "resty.openssl.hmac"
local rand    = require "resty.openssl.rand"
local pae     = require "jitaccess.core.pae"
local canon   = require "jitaccess.core.canon"
local bit     = require "bit"      -- LuaJIT

local bor, bxor = bit.bor, bit.bxor
local byte, sub, char = string.byte, string.sub, string.char
local floor = math.floor
local concat = table.concat

local PROOF_DOMAIN = "jitaccess-v1"
local NONCE_DOMAIN = "jitaccess-nonce-v1"
local NONCE_LEN    = 56

local _M = { _VERSION = "0.1.0", PROOF_DOMAIN = PROOF_DOMAIN, NONCE_DOMAIN = NONCE_DOMAIN }

-- ---- base64url (no padding) ------------------------------------------------

function _M.b64u_encode(s)
  local b = ngx.encode_base64(s, true)          -- no padding, standard alphabet
  return (b:gsub("[+/]", { ["+"] = "-", ["/"] = "_" }))
end

function _M.b64u_decode(s)
  s = (s:gsub("[-_]", { ["-"] = "+", ["_"] = "/" }))
  local rem = #s % 4
  if rem == 2 then s = s .. "=="
  elseif rem == 3 then s = s .. "="
  elseif rem == 1 then return nil end
  return ngx.decode_base64(s)
end

-- ---- primitives ------------------------------------------------------------

local function hmac_sha256(key, data)
  local h, err = hmac.new(key, "sha256")
  if not h then return nil, err end
  local out
  out, err = h:final(data)
  if not out then return nil, err end
  return out
end
_M.hmac_sha256 = hmac_sha256

-- Constant-time equality. Length is not secret (proof/MAC lengths are fixed),
-- so an early length check is safe; the byte loop is branch-free on content.
local function ct_equal(a, b)
  if type(a) ~= "string" or type(b) ~= "string" or #a ~= #b then return false end
  local diff = 0
  for i = 1, #a do
    diff = bor(diff, bxor(byte(a, i), byte(b, i)))
  end
  return diff == 0
end
_M.ct_equal = ct_equal

-- CSPRNG; fails closed (returns nil) on error — callers MUST reject on nil.
function _M.random_bytes(n)
  local b, err = rand.bytes(n)
  if not b or #b ~= n then return nil, err or "rng failed" end
  return b
end

local function be64(n)
  local b = {}
  for i = 8, 1, -1 do
    b[i] = char(n % 256)
    n = floor(n / 256)
  end
  return concat(b)
end

-- ---- proof (the knock) -----------------------------------------------------

function _M.proof_canonical(server_name, kid, nonce_raw)
  return pae.encode({ PROOF_DOMAIN, canon.canon_server_name(server_name), kid, nonce_raw })
end

function _M.build_proof(secret, server_name, kid, nonce_raw)
  return hmac_sha256(secret, _M.proof_canonical(server_name, kid, nonce_raw))
end

function _M.verify_proof(secret, server_name, kid, nonce_raw, tag)
  local expect = _M.build_proof(secret, server_name, kid, nonce_raw)
  if not expect then return false end
  return ct_equal(expect, tag)
end

-- ---- stateless nonce -------------------------------------------------------

local function nonce_mac(nonce_key, ts, rand16, server_name, ip, v6_prefix)
  local ip_c = canon.canon_ip(ip, v6_prefix)
  if not ip_c then return nil, "bad ip" end
  local input = pae.encode({
    NONCE_DOMAIN, be64(ts), rand16,
    canon.canon_server_name(server_name), ip_c,
  })
  return hmac_sha256(nonce_key, input)
end

-- returns nonce (56 bytes, raw) or nil,err
function _M.issue_nonce(nonce_key, ts, rand16, server_name, ip, v6_prefix)
  if #rand16 ~= 16 then return nil, "rand must be 16 bytes" end
  local mac, err = nonce_mac(nonce_key, ts, rand16, server_name, ip, v6_prefix)
  if not mac then return nil, err end
  return be64(ts) .. rand16 .. mac
end

-- returns ok(bool), rand16|nil. Does NOT enforce single-use (see store.NonceStore).
function _M.verify_nonce(nonce_key, nonce, server_name, ip, now, ttl, v6_prefix)
  if type(nonce) ~= "string" or #nonce ~= NONCE_LEN then return false end
  local ts = 0
  for i = 1, 8 do ts = ts * 256 + byte(nonce, i) end
  local rand16 = sub(nonce, 9, 24)
  local mac = sub(nonce, 25, 56)
  local expect = nonce_mac(nonce_key, ts, rand16, server_name, ip, v6_prefix)
  if not expect or not ct_equal(expect, mac) then return false end
  if now < ts or (now - ts) >= ttl then return false end
  return true, rand16
end

return _M

File: core/go/crypto.go
package jitcore

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/binary"
	"errors"
	"strings"
)

const (
	ProofDomain = "jitaccess-v1"       // domain separator for the knock proof
	NonceDomain = "jitaccess-nonce-v1" // domain separator for the nonce MAC
	NonceLen    = 56                   // 8 (ts) + 16 (rand) + 32 (mac)
)

// ---- base64url, no padding -------------------------------------------------

func B64u(b []byte) string { return base64.RawURLEncoding.EncodeToString(b) }

// B64uDecode accepts padded or unpadded input (clients differ; the wire format
// is unpadded).
func B64uDecode(s string) ([]byte, error) {
	return base64.RawURLEncoding.DecodeString(strings.TrimRight(s, "="))
}

// ---- primitives ------------------------------------------------------------

func hmacSHA256(key, data []byte) []byte {
	m := hmac.New(sha256.New, key)
	m.Write(data)
	return m.Sum(nil)
}

// RandomBytes fails closed: an RNG error returns an error rather than degraded
// entropy, and every caller must reject on it (SPEC §7.5).
func RandomBytes(n int) ([]byte, error) {
	b := make([]byte, n)
	if _, err := rand.Read(b); err != nil {
		return nil, err
	}
	return b, nil
}

// ---- proof (the knock) -----------------------------------------------------

// ProofCanonical is the exact byte string the proof MACs over.
func ProofCanonical(serverName, kid string, nonceRaw []byte) []byte {
	return PAE([]byte(ProofDomain), []byte(CanonServerName(serverName)), []byte(kid), nonceRaw)
}

// BuildProof runs identically for a real and a dummy key, so the caller can keep
// timing flat for an unknown kid instead of leaking registry membership
// (PROTOCOL §6 — the equalized-work verifier lives in the adapter).
func BuildProof(secret []byte, serverName, kid string, nonceRaw []byte) []byte {
	return hmacSHA256(secret, ProofCanonical(serverName, kid, nonceRaw))
}

// VerifyProof compares in constant time.
func VerifyProof(secret []byte, serverName, kid string, nonceRaw, tag []byte) bool {
	return hmac.Equal(BuildProof(secret, serverName, kid, nonceRaw), tag)
}

// ---- stateless nonce -------------------------------------------------------

// The nonce carries its own authentication, so /challenge stores nothing: the
// spent-set is bounded by successful knocks rather than by challenge volume and
// cannot be flooded (SECURITY-REVIEW H2/H5). Single-use is enforced separately
// by NonceStore, after the proof verifies.

func nonceMACInput(ts uint64, rand16 []byte, serverName, ip string, v6Prefix int) ([]byte, error) {
	ipc, err := CanonIP(ip, v6Prefix, 32)
	if err != nil {
		return nil, err
	}
	tsb := make([]byte, 8)
	binary.BigEndian.PutUint64(tsb, ts)
	return PAE([]byte(NonceDomain), tsb, rand16, []byte(CanonServerName(serverName)), []byte(ipc)), nil
}

// IssueNonce returns the 56-byte nonce: be64(ts) || rand16 || HMAC.
func IssueNonce(nonceKey []byte, ts uint64, rand16 []byte, serverName, ip string, v6Prefix int) ([]byte, error) {
	if len(rand16) != 16 {
		return nil, errors.New("nonce rand must be 16 bytes")
	}
	in, err := nonceMACInput(ts, rand16, serverName, ip, v6Prefix)
	if err != nil {
		return nil, err
	}
	out := make([]byte, 8, NonceLen)
	binary.BigEndian.PutUint64(out, ts)
	out = append(out, rand16...)
	return append(out, hmacSHA256(nonceKey, in)...), nil
}

// VerifyNonce checks the MAC and the freshness window. It deliberately does NOT
// enforce single-use — that is NonceStore.Claim, called only after the proof
// verifies (verify-then-burn, so a bad proof cannot consume a nonce).
//
// The nonce is bound to the service and the client IP, so one minted for
// another origin or another client will not verify here.
func VerifyNonce(nonceKey, nonce []byte, serverName, ip string, now, ttl int64, v6Prefix int) (bool, []byte) {
	if len(nonce) != NonceLen {
		return false, nil
	}
	ts := binary.BigEndian.Uint64(nonce[:8])
	rand16, mac := nonce[8:24], nonce[24:]
	in, err := nonceMACInput(ts, rand16, serverName, ip, v6Prefix)
	if err != nil {
		return false, nil
	}
	if !hmac.Equal(hmacSHA256(nonceKey, in), mac) {
		return false, nil
	}
	if now < int64(ts) || now-int64(ts) >= ttl {
		return false, nil
	}
	return true, rand16
}
